Magic Link login

kwonye opened the pull request:
Fixes #3959 

To test:

1. Sign in with email you know is WordPress.com email and login by requesting email
2. Sign in with email you know is WordPress.com email and login with password
3. Sign in with username and verify password request is shown and is able to log in
4. Click 'Add self-hosted site` and verify it works

Needs review: @maxme

kwonye commented:
Marking Not Ready for Review until #3971 is merged as we'll need to update how we handle smart lock.

kwonye commented:
Removed A/B testing in this PR and created #3988

kwonye commented:
@maxme all ready to review!

maxme commented:
Seems a bit risky to merge this in `develop` without A/B testing enabled, can I merge it in a feature branch until A/B test lands?

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java:
```diff
@@ -299,18 +317,19 @@ public void onClick(View v) {
 
     private void initSmartLockForPasswords() {
-        if (!isGooglePlayServicesAvailable()) {
+        if (!isSmartLockAvailable()) {
```
Nope we should keep the `isGooglePlayServicesAvailable` test here , `mCredentialsClient` will always be null

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java:
```diff
@@ -299,18 +317,19 @@ public void onClick(View v) {
 
     public void smartLockAutoFill() {
-        if (!isGooglePlayServicesAvailable() || mCredentialsClient == null) {
+        if (!isSmartLockAvailable()) {
```
Same here, we shouldn't use `isSmartLockAvailable` (this method should work when `mCredentialsClient` is disconnected)

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/MagicLinkSignInFragment.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Override
+    protected boolean isSmartLockAvailable() {
```
Is this temporary? Or do you think we should drop smart lock when magic link ship?

kwonye replied:
Definitely temporary. We should not drop smart lock. Reason for this is because A/B testing results between this method and the previous one will be much easier to digest.

kwonye replied:
#3992 added, thanks @maxme

maxme commented:
* First screen: enter your username, switch app (or rotate), username is now blank.
* First screen: "enter key" (hardware keyboard) doesn't work - it's a big deal when you use emulators a lot ;)
* First screen: username field is not focused (virtual keyboard is not opened) when you start the app, so the user have to tap the username field.
* "Get a link" screen: title is tappable, but after 10 taps, no easter eggs :(
* "Get a link" screen: tap help button does nothing :sadpanda:
* "Get a link" screen: "Enter your password instead", I wonder if we should style that button the same way we do for "Send link" (we have a lot of space on this screen) - and even if we want users to prefer the "send link" button, it's important enough to have it equally represented on that screen.
* "Check your email" screen: improve wording maybe? "Check your email"/"Open mail" - what about "Open your mailbox" or "Launch your email app".
* "Check your email" screen: no help button here?
* All these screens: on activity restart all changes are lost and the user is back on the first screen.
* All these screens: fragment stack is probably wrong, the app does really weird things when you tap the back button.

maxme commented:
App crashed when I tried to add a new site from the Site picker:

```
04-20 16:19:21.054 6301-6301/org.wordpress.android.beta E/AndroidRuntime: FATAL EXCEPTION: main
Process: org.wordpress.android.beta, PID: 6301
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.wordpress.android.beta/org.wordpress.android.ui.accounts.SignInActivity}: java.lang.Nulvirtual method 'void android.widget.RelativeLayout.setVisibility(int)' on a null object reference
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2298)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2360)
    at android.app.ActivityThread.access$800(ActivityThread.java:144)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1278)
    at android.os.Handler.dispatchMessage(Handler.java:102)
    at android.os.Looper.loop(Looper.java:135)
    at android.app.ActivityThread.main(ActivityThread.java:5221)
    at java.lang.reflect.Method.invoke(Native Method)
    at java.lang.reflect.Method.invoke(Method.java:372)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
 Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.RelativeLayout.setVisibility(int)' on a null object reference
    at org.wordpress.android.ui.accounts.SignInFragment.forceSelfHostedMode(SignInFragment.java:248)
    at org.wordpress.android.ui.accounts.SignInActivity.actionMode(SignInActivity.java:67)
    at org.wordpress.android.ui.accounts.SignInActivity.onCreate(SignInActivity.java:47)
    at android.app.Activity.performCreate(Activity.java:5933)
    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1105)
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2251)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2360) 
    at android.app.ActivityThread.access$800(ActivityThread.java:144) 
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1278) 
    at android.os.Handler.dispatchMessage(Handler.java:102) 
    at android.os.Looper.loop(Looper.java:135) 
    at android.app.ActivityThread.main(ActivityThread.java:5221) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at java.lang.reflect.Method.invoke(Method.java:372) 
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694) 
```

maxme commented:
I'm not sure what's happening but on 4.1.1 geny emulator, when trying to login with username + password (wpcom or self hosted), after hitting the "Sign in" button, the loading indicator never stops.

maxme commented:
Very weird behavior:

1. From a clean app state
1. Sign in with a self hosted
1. Tap "Connect to WordPress.com" on the Me tab
1. Use the magic link feature
1. Open email app, tap the magic link, that reopens the app "Logging in..."
1. And then back on the email app
1. Reopening the app shows a "Check your email" screen
1. Tapping back goes to "Send email"
1. Killing the app / restarting - I'm logged in the wpcom account

maxme commented:
Small issue:

1. From a clean app state
1. Sign in with a self hosted
1. Tap "Connect to WordPress.com" on the Me tab - we shouldn't have the "Add self hosted site" button here.

kwonye commented:
> First screen: enter your username, switch app (or rotate), username is now blank.
> All these screens: on activity restart all changes are lost and the user is back on the first screen.
> All these screens: fragment stack is probably wrong, the app does really weird things when you tap the back button.
> Very weird experience
> I'm not sure what's happening but on 4.1.1 geny emulator, when trying to login with username + password (wpcom or self hosted), after hitting the "Sign in" button, the loading indicator never stops.

I found the root cause for all the "weird" issues: I mistakenly created multiple instances of `MagicLinkSignInFragment`. I've reimplemented how `MagicLinkSignInActivity` handles retrieving the `MagicLinkSignInFragment` (Uses the FragmentManager instead of relying on a member variable). All the "weird" behavior is non-reproducable with the latest changes.

> First screen: "enter key" (hardware keyboard) doesn't work - it's a big deal when you use emulators a lot ;)

Addressed in 7c63a08

> First screen: username field is not focused (virtual keyboard is not opened) when you start the app, so the user have to tap the username field.

That was intentional. I think it's more "courteous" and consistent with other apps. If you feel strongly about this however, I'll defer to your judgement and change it back.

> "Get a link" screen: title is tappable, but after 10 taps, no easter eggs :(

You didn't try 11? ;)

> "Get a link" screen: tap help button does nothing :sadpanda:
> "Check your email" screen: no help button here?

Addressed in 353bf9f

> "Get a link" screen: "Enter your password instead", I wonder if we should style that button the same way we do for "Send link" (we have a lot of space on this screen) - and even if we want users to prefer the "send link" button, it's important enough to have it equally represented on that screen.
> "Check your email" screen: improve wording maybe? "Check your email"/"Open mail" - what about "Open your mailbox" or "Launch your email app".

For V1 I'd like to keep it consistent with both iOS (generally) and @mattmiklic's design. That said, this would be another great A/B testing candidate (A/B TEST ALL THE THINGS!)

Thanks for the great review! 🙇

maxme commented:
> > "Check your email" screen: improve wording maybe? "Check your email"/"Open mail" - what about "Open your mailbox" or "Launch your email app".

> For V1 I'd like to keep it consistent with both iOS (generally)

Right, but on iOS, the email app is called "Mail", there is no such app in Android.

kwonye commented:
> Right, but on iOS, the email app is called "Mail", there is no such app in Android.

Can't deny that. Addressed in 33551ff

maxme commented:
* I just noticed you can tap several time the "Next" button: https://cloudup.com/cZriRtIYPnb
* I can repro this https://github.com/wordpress-mobile/WordPress-Android/pull/3973#issuecomment-212450130 and this https://github.com/wordpress-mobile/WordPress-Android/pull/3973#issuecomment-212448301
* Enter an username, tap next, password field shows up, rotate the screen: password field is hidden.

kwonye commented:
> I'm not sure what's happening but on 4.1.1 geny emulator, when trying to login with username + password (wpcom or self hosted), after hitting the "Sign in" button, the loading indicator never stops.

Hmm, that's weird because that is directly calling `super.signIn()`. Do you mind trying it on the original SignInActivity? I don't have a genymotion emulator, but I was able to log in fine with the Google 4.1 emulator

kwonye commented:
> I just noticed you can tap several time the "Next" button: https://cloudup.com/cZriRtIYPnb

Addressed in 7684adc

> Enter an username, tap next, password field shows up, rotate the screen: password field is hidden.

Addressed in 987a2f4

> [weird email issue](https://github.com/wordpress-mobile/WordPress-Android/pull/3973#issuecomment-212450130)

Addressed in a57ab46

kwonye commented:
@maxme thanks for the latest pass. One note: I set MagicLinkSignInActivity to false (fb46211) as the default, so you'll need to set that back to true 👍

maxme commented:
:shipit: 
Note: the feature should be enabled on the beta only first
